components/datetime/src/format.rs

```diff
@@ -83,112 +88,199 @@ where
-                    format_number(w, date_time.second().into(), field.length)?
                 }
-                FieldSymbol::DayPeriod(period) => {
-                    let symbol = data.get_symbol_for_day_period(
-                        period,
-                        field.length,
-                        date_time.hour(),
-                        is_top_of_hour(&pattern, date_time),
-                    );
-                    w.write_str(symbol)?
+                fields::Hour::H23 => h,
+                fields::Hour::H24 => {
+                    if h == 0 {
+                        24
+                    } else {
+                        h
+                    }
                 }
-            },
-            PatternItem::Literal(l) => w.write_str(&l)?,
+            };
+            format_number(w, value, &field.length)?
         }
-    }
+        FieldSymbol::Minute => format_number(
+            w,
+            usize::from(
+                date_time
+                    .date_time()
+                    .minute()
+                    .ok_or(Error::MissingInputField)?,
+            ) as isize,
+            &field.length,
+        )?,
+        FieldSymbol::Second(..) => format_number(
+            w,
+            usize::from(
+                date_time
+                    .date_time()
+                    .second()
+                    .ok_or(Error::MissingInputField)?,
+            ) as isize,
+            &field.length,
+        )?,
+        FieldSymbol::DayPeriod(period) => {
+            let symbol = data.get_symbol_for_day_period(
+                period,
+                field.length,
+                date_time
+                    .date_time()
+                    .hour()
+                    .ok_or(Error::MissingInputField)?,
+                arithmetic::is_top_of_hour(
+                    &pattern,
+                    date_time.date_time().minute().map(u8::from).unwrap_or(0),
+                    date_time.date_time().second().map(u8::from).unwrap_or(0),
+                ),
+            );
+            w.write_str(symbol)?
+        }
+    };
     Ok(())
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
+    #[test]
+    #[cfg(feature = "provider_serde")]
+    fn test_basic() {
+        use crate::mock::MockDateTime;
+        use icu_provider::prelude::*;
+        let provider = icu_testdata::get_provider();
+        let data = provider
+            .load_payload(&DataRequest {
+                resource_path: ResourcePath {
+                    key: provider::key::GREGORY_V1,
+                    options: ResourceOptions {
+                        variant: None,
+                        langid: Some("en".parse().unwrap()),
+                    },
+                },
+            })
+            .unwrap()
+            .take_payload()
+            .unwrap();
+        let pattern = crate::pattern::Pattern::from_bytes("MMM").unwrap();
+        let date_time = MockDateTime::try_new(2020, 8, 1, 12, 34, 28).unwrap();
```

Comment:
I'm strongly against overuse of the builder pattern here. Builder pattern is the right choice if the state of the struct makes sense in between various calls. In this case the `MockDateTime` makes no sense with half of the field initialized, and using a builder pattern for that is imho a bad API design model (which I've seen way too often).

Instead, we could use a constructor options struct:

```rust
MockDateTime::try_new(MockDateTimeOptions {
    year: 2020,
    month: 8,
    day: 1,
    hour: 12,
    minute: 34,
    second: 28
});
```

which is analogous to builder pattern in the expressiveness but doesn't bring the problem of half-initialized states.

In this particular case tho, I think that a long list of arguments is fine, since it's a mock and we don't intend it to be the final user facing pattern in the long run.

Reply:
I admit to being guilty of "overusing" the builder pattern for only the sake of readability. I just find it to be incredibly readable and expressive. 

But I don't disagree with what you just said.

Reply:
The issue with the builder pattern mid-state is two fold:

1) You can, and by the law of nature since you can you will, end up with a half-initialized structs. You can of course end up with `.build()` method that fails if the struct is not ready, but that means you moved architectural error reporting from compile time to run time at a loss.
2) When you change the list of fields required, you have no way to inform your users that they now need to call one more method on the builder pattern, except of, again, failing at `.build()` method, which is runtime.

Reply:
All valid points. No arguments from me.

Reply:
I'm not interested in rewriting MockDateTime in this PR.  I think it should be thrown out in the dumpster because it makes several really bad assumptions, not to mention the fact that it only works in proleptic Gregorian.  When we come up with a real datetime struct, we can engage these design discussions.